### ISSUE TITLE
Feature request: split comments, not just code

### ISSUE THREAD

**Issue by cpitclaudel** [ejgallego/coq-serapi#200]
Hi Emilio,

Hope you're doing OK despite the crazy times.
In the project I'm workig on, it would be very useful to be able to get comments boundaries in addition to code fragments; like this:

```
("query0" (Add () "Goal True.\n  (* R1 *)  (** D1 **) idtac. (* R2 *)"))
(Answer query0 Ack)
(Answer query0(Added 2((fname ToplevelInput)(line_nb 1)(bol_pos 0)(line_nb_last 1)(bol_pos_last 0)(bp 0)(ep 10))NewTip))
(Answer query0(Added 3((fname ToplevelInput)(line_nb 2)(bol_pos 11)(line_nb_last 2)(bol_pos_last 11)(bp 34)(ep 40))NewTip))
(Answer query0 Completed)
```

It would be great if instead of two spans (`0-10`, `34-40`), SerAPI could return 5: `Goal True.`, `(* R1 *)`, `(** D1 **)`, `idtac.`, `(* R2 *)`.

Currently I have a mini-parser that I use after passing things through SerAPI, to detect comment boundaries.

WDYT?

Thanks!

**Comment by ejgallego** [comment]
Hi Clément, thanks, I hope you folks are doing well too!

This is pretty similar to #191 , and it's been on the table for a while, I just lack the cycles right now as this would involve modifying Coq's parser. I will provide some workaround for 8.12 tho, so at least the Python (client) side can recover comments from files and slice them back.

**Comment by ejgallego** [comment]
Indeed, what we can provide for now is a wrapper to 8.12's:
```
    val get_comments : t -> ((int * int) * string) list
```

**Comment by cpitclaudel** [comment]
That would be very nice, indeed.

**Comment by ejgallego** [comment]
> That would be very nice, indeed

Still this defers the insertion of comments to the clients which is not ideal, the problem is that Coq doesn't recognize comments as part of the document, so they don't have a `statement_id`, but if we could do this SerAPI-side it would be great, maybe we could add a call that would output the document decorated somehow.

**Comment by ejgallego** [comment]
#205 should help with this, will be included in the next release. Still far from what you would like [likely impossible to solve properly unless we introduce #5 first).

For now, I'm not closing but postponing further improvements.

**Comment by cpitclaudel** [comment]
Thanks a lot, sorry for the long delay in my answer.  This looks good.  I think your link to #5 is quite relevant too, indeed; for now for that I have a comment parser written in Python.

**Comment by ejgallego** [comment]
In the coq-lsp model, we support hybrid documents and the workflow for clients should not have this problem.

See also #318

**Comment by cpitclaudel** [comment]
Can you clarify how hybrid documents work? / Point to some docs?

**Comment by ejgallego** [comment]
Hi @cpitclaudel , hybrid documents work by sending the full markdown (or json) file to the server, so the clients still own the whole buffer.

That's actually not so different with SerAPI if used in full document mode. So for the comments use-case I think the hybrid nature is not relevant actually.

The important difference IMO is that the interface is position based, and that you don't really have a notion of "document split" anymore, but more of range inclusion.

Then, given a position in the document, it can belong to one or more range, ordered by inclusion. But these ranges are in direct mapping with the document, so the clients know that any gap in the range tree, corresponds to contents in the doc text itself. In this sense, that's just a box model which would correspond to printing.

Indeed to do some fancy printing with inline comments client can just see the Ast boxes, so in that sense the situation hasn't changed. I don't know how to put the comments in the Ast directly without complexifying it a lot. Maybe lexing would the right level for this?

There is no API for doing such a range query, but can be added easily, once it is defined.

Hybrid does help in some examples, in software foundations you have:
```coq
Inductive :=
| A of int (* this is A *)
| B or int (* this is B *)
.
```                  
which in JsCoq was very problematic as you couldn't split that into sentences where the comments are turned into html by coqdoc, but now you can map that to:
```markdown
\```coq
Inductive :=
| A of int
\```
this is A
\```coq
 B or int
.
\```
this is B.
```
which seems better.

**Comment by ejgallego** [comment]
To be more clear, while comments are not part of the Ast, they are part of the document in the sense that for any position, if you query for its info, you may get a comment range.

**Comment by cpitclaudel** [comment]
> hybrid documents work by sending the full markdown (or json) file to the server, so the clients still own the whole buffer.

But here what I have is a Coq document, and I'm trying to split it up into sentences + comments for processing; that wasn't supported in SerAPI (only the sentences, but not comment boundaries).  Is it supported by LSP?

> which in JsCoq was very problematic as you couldn't split that into sentences where the comments are turned into html by coqdoc, 

I'm a bit confused.  The way Alectryon works with LSP servers today is that it collects the code, concatenates everything, and then sends that to the prover.  So from the prover's perspective the two examples are exactly the same.

**Comment by ejgallego** [comment]
> But here what I have is a Coq document, and I'm trying to split it up into sentences + comments for processing; that wasn't supported in SerAPI (only the sentences, but not comment boundaries). Is it supported by LSP?

What I mean is that splitting doesn't really make a lot of sense, at least with the current definition.

What is the splitting of `Definition (* foo*) a := (* bar *) forall (x : (nat * (* a type *) nat), True.` ?

> I'm a bit confused. The way Alectryon works with LSP servers today is that it collects the code, concatenates everything, and then sends that to the prover. So from the prover's perspective the two examples are exactly the same.

Yes, they are the same, but for IDEs with sentence models the splitting was a pain as you would have two editor boxes. But indeed maybe this example is not relevant.

**Comment by cpitclaudel** [comment]
> What I mean is that splitting doesn't really make a lot of sense, at least with the current definition of splitting.

I think the current definition is pretty well established: sentences are the smallest unit that Coq can process (which happens to end in a period).  The output of coqc -time, in other words.

> What is the splitting of …

That a single sentence, no?

(I'm just trying to recover enough functionality to rebuild alectryon on top of LSP; one huge problem we had with Lean was that the API only let you ask "what's the goal at this point", and not "which part of the document does this goal apply to?")

**Comment by ejgallego** [comment]
> That a single sentence, no?

Yes, sorry, I was asking about splitting the comments, I'm not sure I understand how would you like the comments to be split.

If you run coq-lsp 0.1.4 (in 0.1.5 this info is only available in debug builds, I'll add an option so it can be displayed again) you can see how the sentences are split using hover. This is also available in https://github.com/ejgallego/coq-lsp/pull/351 .

So in Coq case, this document:
```coq
(* foo *)
Definition bar := moo.
(* bar *)
Definition woo := ups.
```
are two sentences, and indeed they include the comments, so the ranges are `[(0,0)-(1,23); (2,0)-(3,23)]`,

Comments are included in the sentences, viewed as text, but the Ast can't, as for example this is not possible to capture:
```
Set (* foo *) Printing All.
```
can't be properly captured in the Ast. But actually, the way Flèche organizes stuff is as an "overlay" of different meta-data. The only data that is guaranteed to happen is a text range, this is because a document can fail to parse.

If the document parses, then an Ast is present, and then we refine the ast to generate sub-ranges corresponding to the identifiers. Comment are the same but they live in a pure textual layer, so for the above example you get ranges:
```
text : (0,0)-(0,34)
comments : [(0,4)-(0,14)]
ast: (0,0)-(0,34)
ids : []
```
etc... so the splitting we do is really a kind of nesting, and there are different structures that are not really comparable outside the positional view.

Keep in mind that little of this is exposed in the LSP protocol, but it is fairly easy to do so if we want.

Also note that there are sentences such as bullets or braces that don't end in a dot, and in the tactic engine we can process `tac1; tac2` in a finer way (as done by the ltac debugger). So the reality is slightly more subtle.

We don't store the lexing tokenization as Coq lacks the API to do so, but we could actually do so.

> (I'm just trying to recover enough functionality to rebuild alectryon on top of LSP; one huge problem we had with Lean was that the API only let you ask "what's the goal at this point", and not "which part of the document does this goal apply to?")

I think the `coq/getDocument` request should work for you indeed, but note different modes that we plan to enable in coq-lsp such as the one i just mentioned where `tac1; tac2` indeed is single sentence, but coq-lsp provides different goal info for the position of `tac1` and `tac2`.

Does that make sense?
